### PR TITLE
[DSV4] perf: Make FP8 quant output tensor contiguous

### LINKED ISSUE #29038: [BUG] DeepSeek-V4 FP8 wo_a DeepGEMM path can lower accuracy on Blackwell
## Summary

SGLang can show lower benchmark accuracy than vLLM for DeepSeek-V4-Flash W4A8 on Blackwell / SM100. This issue uses GPQA Diamond as the concrete reproducer, but the problem is not GPQA-specific.

The suspected cause is the FP8 `wo_a` activation scale layout passed to `deep_gemm.fp8_einsum` in `python/sglang/srt/models/deepseek_v4.py`. Since this path can change model logits during inference, the bug causes low accuracy across downstream evaluations. We have confirmed this on GPQA, AIME25, and SWE-Bench; after applying the layout fix, accuracy recovered on these evaluations.

Fix PR: https://github.com/sgl-project/sglang/pull/29036

## Suspected Root Cause

The current FP8 `wo_a` path quantizes `o: [T, G, D]` by flattening it into `[T * G, D]`, then views the packed UE8M0 scales back as `[T, G, -1]`.

On the SM100 DeepGEMM FP8 einsum path, this can mismatch activation scales with groups and cause early logit divergence. PR #29036 changes the quantization/scale layout to preserve group-major semantics.

## Impact

This is an inference correctness issue, not only a GPQA-specific regression. The incorrect FP8 `wo_a` scale layout introduces logit divergence in the DeepGEMM path, so benchmarks and workloads that depend on stable reasoning, code generation, or action generation can be affected.

GPQA is used here because it is a small and repeatable benchmark where we observed a clear accuracy gap and recovery after the layout fix. We have also reproduced low-accuracy symptoms on AIME25 and SWE-Bench, and after applying the same layout fix, accuracy recovered on those evaluations as well.

Confirmed affected evaluations include:

- GPQA
- AIME25
- SWE-Bench

Other math, reasoning, coding, and agentic benchmarks may also be affected by the same inference correctness issue.

## Environment

- Model: DeepSeek-V4-Flash W4A8
- Hardware: Blackwell / SM100
- SGLang baseline: `v0.5.12.post1`
- Reproducer dataset: GPQA Diamond, 198 examples

## Server Command

Baseline and patched SGLang used the same server arguments. The patched build used `PYTHONPATH=/path/to/sglang-pr/python:$PYTHONPATH`.

```bash
export SGLANG_OPT_DEEPGEMM_MEGA_MOE_NUM_MAX_TOKENS_PER_RANK=8320

sglang serve \
  --trust-remote-code \
  --model-path /path/to/DeepSeek-V4-Flash \
  --tp 2 \
  --dp 1 \
  --enable-dp-attention \
  --moe-a2a-backend megamoe \
  --host 0.0.0.0 \
  --port 31000
```

## Reproducer Test Command (GPQA)

Full GPQA was run three times for each build/server:

```bash
sgl-eval run gpqa \
  --model /path/to/DeepSeek-V4-Flash \
  --api-key EMPTY \
  --base-url http://localhost:31000/v1 \
  --from-dataset /path/to/gpqa_diamond.jsonl \
  --num-threads 16 \
  --n-repeats 1 \
  --max-tokens 200000 \
  --temperature 0.0 \
  --top-p 1.0 \
  --thinking
```

## GPQA Reproducer Results

The following GPQA numbers are one concrete reproducer for the broader inference-accuracy issue.

| System | GPQA scores from 3 runs | Mean |
| --- | --- | ---: |
| vLLM baseline | 0.863636, 0.858586, 0.873737 | 0.865320 |
| SGLang before fix | 0.843434, 0.828283, 0.833333 | 0.835017 |
| SGLang after PR #29036 | 0.883838, 0.863636, 0.863636 | 0.870370 |

After the layout fix, SGLang recovered to the vLLM baseline range on this GPQA reproducer. We have also confirmed recovery on AIME25 and SWE-Bench after the same fix.

### PR DESCRIPTION
**Pull request by mattteochen**
## Motivation

This PR is an alternative way to achieve https://github.com/sgl-project/sglang/pull/27680.

Deep gemm FP8 einsum silently falls back to an unoptimized path if the input tensor is not contiguous. We can make the `per_token_group_quant_8bit_v2` to emit the scale layout needed by deep gemm.

This also implicitly fixes https://github.com/sgl-project/sglang/issues/29038. The previous DSV4 `wo_a` FP8 DeepGEMM path flattened `o: [T, G, D]` into `[T * G, D]` and then viewed the scales back as `[T, G, -1]`. On the SM100 FP8 einsum path, scales are consumed as per-group slabs, so flattening across `T * G` can associate a token's FP8 codes with another group's scale. This PR quantizes `o` as `[T, G, D]` and emits the scale tensor in DeepGEMM's expected outer/group-major layout, preserving the `scale[t, g]` to `o[t, g, :]` association.

image

image

## Modifications

- Extended the fp8 quant jit v2 kernel to emit dsv4 fp8 einsum required layout.
- Added related tests.
- Updated dsv4 model.

## Accuracy Tests

```bash
== gpqa ==
198 examples x 16 repeats  |  6980.5s  |  4956 tok/s  |  34.6M tokens

* pass@1[avg-of-16]  =  88.04% +/- 1.56% (SEM 0.39%)
  pass@16            =  96.46%
  majority@16        =  90.15%
  no_answer          =  0.06%
  stop_rate          =  100.00%
  truncated_rate     =  0.00%
  error_rate         =  0.00%
```

```bash 
== gsm8k ==
1319 examples (single-shot)  |  156.8s  |  1125 tok/s  |  176K tokens

* score           =  96.29%
  stop_rate       =  100.00%
  truncated_rate  =  0.00%
  error_rate      =  0.00%
```

## Speed Tests and Profiling

Node: 8XB300 + DSV4-Pro

Pareto for 1k1k, no MTP:
```bash
series                            conc       TPS/user (sp)    out TPS/GPU (sp)  total TPS/GPU (sp)        TTFT ms (sp)
----------------------------------------------------------------------------------------------------------------------
base                                 4      81.90 ( 1.00x)      39.63 ( 1.00x)      79.78 ( 1.00x)     267.20 ( 1.00x)
base                                 8      70.18 ( 1.00x)      67.67 ( 1.00x)     135.75 ( 1.00x)     244.00 ( 1.00x)
base                                16      55.89 ( 1.00x)     109.25 ( 1.00x)     216.89 ( 1.00x)     244.17 ( 1.00x)
base                                32      41.89 ( 1.00x)     164.00 ( 1.00x)     328.88 ( 1.00x)     245.75 ( 1.00x)
base                                64      28.97 ( 1.00x)     231.01 ( 1.00x)     462.65 ( 1.00x)     250.78 ( 1.00x)
opt                                  4      85.37 ( 1.04x)      41.01 ( 1.03x)      82.54 ( 1.03x)     235.97 ( 1.13x)
base                                64      28.96 ( 1.00x)     230.75 ( 1.00x)     462.12 ( 1.00x)     247.41 ( 1.00x
opt                                  8      73.05 ( 1.04x)      70.42 ( 1.04x)     141.26 ( 1.04x)     234.98 ( 1.04x)
opt                                 16      57.76 ( 1.03x)     112.98 ( 1.03x)     224.30 ( 1.03x)     236.35 ( 1.03x)
opt                                 32      43.05 ( 1.03x)     168.95 ( 1.03x)     338.82 ( 1.03x)     239.09 ( 1.03x)
opt                                 64      29.63 ( 1.02x)     236.42 ( 1.02x)     473.48 ( 1.02x)     241.48 ( 1.02x)
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #28783296006](https://github.com/sgl-project/sglang/actions/runs/28783296006)
Latest PR Test (Extra): :x: [Run #28783295767](https://github.com/sgl-project/sglang/actions/runs/28783295767)

### CODE REVIEW

**Comment by mmangkad** [comment]
This plus #26765/#26766 together is basically the AOT version of the all-in-one JIT path I had in #27818, linking it here for context

cc @b8zhong

**Comment by nvpohanh** [comment]
@Fridge003 @b8zhong Could you assign to someone to review this? Thanks!

File: python/sglang/srt/layers/quantization/fp8_kernel.py [hunk 1]
```diff
@@ -462,7 +462,49 @@ def create_per_token_group_quant_fp8_output_scale(
     column_major_scales: bool,
     scale_tma_aligned: bool,
     scale_ue8m0: bool,
+    scale_outer_major: bool = False,
 ):
+    # TODO: replace the scale layout booleans with a single enum before adding
```
Comment (Fridge003) [inline python/sglang/srt/layers/quantization/fp8_kernel.py]: Maybe add some comments explaining the function of `scale_outer_major`

File: python/sglang/srt/layers/quantization/fp8_kernel.py [hunk 2]
```diff
@@ -462,7 +462,49 @@ def create_per_token_group_quant_fp8_output_scale(
...
+    #   ue8m0 + outer_major + tma_aligned       -> per-outer-slice packed int32 UE8M0 (DeepGEMM native)
+    # scale_tma_aligned in any other combination is rejected below rather than
+    # silently ignored.
+    if scale_outer_major:
+        assert scale_ue8m0
+        assert not column_major_scales
+        assert len(x_shape) == 3
+        *x_batch, x_s_mn, x_s_outer, x_q_k = x_shape
+        if scale_tma_aligned:
+            # DeepGEMM's native SM100 SFA layout, one packed slab per outer
```
Comment (Fridge003) [inline python/sglang/srt/layers/quantization/fp8_kernel.py]: Remove the comments here

File: python/sglang/srt/models/deepseek_v4.py [hunk 3]
```diff
@@ -1085,14 +1087,19 @@ def forward(
...
             o_fp8, o_s = sglang_per_token_group_quant_fp8(
-                o.reshape(T * G, D).contiguous(),
+                o.contiguous(),
                 group_size=128,
                 scale_ue8m0=True,
+                scale_outer_major=True,
+                # Packed int32 UE8M0 scales are consumed natively only by
+                # DeepGEMM's SM100 runtime; other archs (e.g. SM90) need the
+                # fp32 power-of-two scales deep_gemm transforms internally.
+                scale_tma_aligned=_is_sm100_supported,
```
Comment (Fridge003) [inline python/sglang/srt/models/deepseek_v4.py]: Instead of adding the `scale_outer_major` argument for `sglang_per_token_group_quant_fp8`, can we create a new kernel specified for this usage (a fp8 quant kernel that supports scale_outer_major, as in https://github.com/sgl-project/sglang/pull/29036).

So we can avoid all those if-else logics in `per_token_group_quant_8bit_v2` kernel files.

Comment (Fridge003) [inline python/sglang/srt/models/deepseek_v4.py]: Why are we setting scale_tma_aligned to True? If it's not necessary can we don't modify this argument in this PR, and potentially open another PR for fixing `scale_tma_aligned`

**Review by Fridge003 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/models/deepseek_v4.py (hunk 3 above)
Comment (mattteochen) [reply python/sglang/srt/models/deepseek_v4.py]: I can do a follow up PR for this, it was needed to remove deep_gemm einsum TMA scaling kernel to be executed.

**Comment by mattteochen** [comment]
On B300 with eval commands from https://github.com/sgl-project/sglang/pull/29036:
```bash
== gpqa ==
198 examples (single-shot)  |  1796.4s  |  1146 tok/s  |  2.1M tokens

* score           =  89.39%
  stop_rate       =  100.00%
  truncated_rate  =  0.00%
  error_rate      =  0.00%
```

**Comment by mattteochen** [comment]
Updated Pareto 1k1k 8xB300 DSV4-pro low latency config no MTP:
image
```bash
series                            conc       TPS/user (sp)    out TPS/GPU (sp)  total TPS/GPU (sp)        TTFT ms (sp)
----------------------------------------------------------------------------------------------------------------------
base                                 4      85.32 ( 1.00x)      41.72 ( 1.00x)      83.98 ( 1.00x)      87.86 ( 1.00x)
base                                 8      77.83 ( 1.00x)      75.20 ( 1.00x)     150.85 ( 1.00x)      89.23 ( 1.00x)
base                                16      65.91 ( 1.00x)     128.43 ( 1.00x)     254.98 ( 1.00x)      92.10 ( 1.00x)
base                                32      52.37 ( 1.00x)     202.78 ( 1.00x)     406.65 ( 1.00x)      97.33 ( 1.00x)
base                                64      39.37 ( 1.00x)     307.24 ( 1.00x)     615.31 ( 1.00x)     103.96 ( 1.00x)
opt                                  4      89.95 ( 1.05x)      43.95 ( 1.05x)      88.48 ( 1.05x)      85.59 ( 1.03x)
opt                                  8      82.34 ( 1.06x)      79.50 ( 1.06x)     159.47 ( 1.06x)      86.41 ( 1.03x)
opt                                 16      69.19 ( 1.05x)     134.75 ( 1.05x)     267.51 ( 1.05x)      89.23 ( 1.03x)
opt                                 32      54.50 ( 1.04x)     211.11 ( 1.04x)     423.36 ( 1.04x)      94.28 ( 1.03x)
opt                                 64      40.45 ( 1.03x)     313.63 ( 1.02x)     628.10 ( 1.02x)     101.19 ( 1.03x)
```

**Review by Fridge003 — APPROVED** [review]

**Comment by Fridge003** [comment]
/rerun-test test/registered/models_e2e/test_deepseek_v4_flash_fp4_h200.py test/registered/models_e2e/test_deepseek_v4_flash_fp8_h200.py test/registered/models_e2e/test_deepseek_v4_flash_fp4_b200.py test/registered/models_e2e/test_deepseek_v4_flash_fp4_megamoe_b200.py test/registered/cp/test_deepseek_v4_flash_fp4_b200_cp.py test/registered/disaggregation/test_disaggregation_dsv4.py

**Comment by Fridge003** [comment]
/rerun-test test/registered/gb300/test_deepseek_v4_pro_fp4.py
